python/ray/data/aggregate.py

```diff
@@ -544,27 +572,113 @@ def _is_null(a: Optional[AggType]) -> bool:
     return a is None or is_nan(a)
 
 
-def _null_safe_merge(
-    merge: Callable[[AggType, AggType], AggType],
+def _null_safe_zero_factory(zero_factory, ignore_nulls: bool):
+    """NOTE: PLEASE READ CAREFULLY BEFORE CHANGING
+
+    Null-safe zero factory is crucial for implementing proper aggregation
+    protocol (monoid) w/o the need for additional containers.
+
+    Main hurdle for implementing proper aggregation semantic is to be able to encode
+    semantic of an "empty accumulator" and be able to tell it from the case when
+    accumulator is actually holding null value:
+
+        - Empty container can be overridden with any value
+        - Container holding null can't be overridden if ignore_nulls=False
+
+    However, it's possible for us to exploit asymmetry in cases of ignore_nulls being
+    True or False:
+
+        - Case of ignore_nulls=False entails that if there's any "null" in the sequence,
+         aggregation is undefined and correspondingly expected to return null
+
+        - Case of ignore_nulls=True in turn, entails that if aggregation returns "null"
+        if and only if the sequence does NOT have any non-null value
+
+    Therefore, we apply this difference in semantic to zero-factory to make sure that
+    our aggregation protocol is adherent to that definition:
+
+        - If ignore_nulls=True, zero-factory returns null, therefore encoding empty
+        container
+        - If ignore_nulls=False, couldn't return null as aggregation will incorrectly
+        prioritize it, and instead it returns true zero value for the aggregation
+        (ie 0 for count/sum, -inf for max, etc).
+    """
+
+    def _safe_zero_factory(_):
+        if ignore_nulls:
+            return None
+
+        return zero_factory()
+
+    return _safe_zero_factory
+
+
+def _null_safe_aggregate(
+    aggregate: Callable[[Block], AggType],
     ignore_nulls: bool,
+) -> Callable[[Block], Optional[AggType]]:
+    def _safe_aggregate(block: Block) -> Optional[AggType]:
+        result = aggregate(block)
+        # NOTE: If `ignore_nulls=True`, aggregation will only be returning
+        #       null if the block does NOT contain any non-null elements
+        if _is_null(result) and ignore_nulls:
+            return None
+
+        return result
+
+    return _safe_aggregate
+
+
+def _null_safe_finalize(
+    finalize: Callable[[AggType], AggType]
+) -> Callable[[Optional[AggType]], AggType]:
+    def _safe_finalize(acc: Optional[AggType]) -> AggType:
+        # If accumulator container is not null, finalize.
+        # Otherwise, return as is.
+        return acc if _is_null(acc) else finalize(acc)
+
+    return _safe_finalize
+
+
+def _null_safe_combine(
```

Comment:
Can you also add some unit test for these wrappers?

Reply:
These are all tested explicitly in `test_null_safe_aggregation_protocol`